exporters/autoexport/registry.go

```diff
@@ -0,0 +1,138 @@
+// implementations that is safe for concurrent use by multiple goroutines
+// without additional locking or coordination.
+type registry struct {
+	mu    sync.Mutex
+	names map[string]trace.SpanExporter
+}
+
+var (
+	// envRegistry is the index of all supported exporter types
+	// and their mapping to a SpanExporter.
+	envRegistry = &registry{
+		names: map[string]trace.SpanExporter{},
+	}
+
+	// errUnknownExpoter is returned when an unknown exporter name is used in
+	// the OTEL_*_EXPORTER environment variables.
+	errUnknownExpoter = errors.New("unknown exporter")
+
+	// errInvalidOtlpProtocol is returned when an invalid protocol is used in
+	// the OTEL_EXPORTER_OTLP_PROTOCOL environment variable.
+	errInvalidOtlpProtocol = errors.New("invalid OTLP protocol - should be one of ['grpc', 'http/protobuf']")
+)
+
+// load returns the value stored in the registry index for a key, or nil if no
+// value is present. The ok result indicates whether value was found in the
+// index.
+func (r *registry) load(key string) (p trace.SpanExporter, ok bool) {
+	r.mu.Lock()
+	p, ok = r.names[key]
+	r.mu.Unlock()
+	return p, ok
+}
+
+var errDupReg = errors.New("duplicate registration")
+
+// store sets the value for a key if is not already in the registry. errDupReg
+// is returned if the registry already contains key.
+func (r *registry) store(key string, value trace.SpanExporter) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if r.names == nil {
+		r.names = map[string]trace.SpanExporter{key: value}
+		return nil
+	}
+	if _, ok := r.names[key]; ok {
+		return fmt.Errorf("%w: %q", errDupReg, key)
+	}
+	r.names[key] = value
+	return nil
+}
+
+// drop removes key from the registry if it exists, otherwise nothing.
+func (r *registry) drop(key string) {
+	r.mu.Lock()
+	delete(r.names, key)
+	r.mu.Unlock()
+}
+
+// RegisterSpanExporter sets the SpanExporter e to be used when the
+// OTEL_TRACES_EXPORTERS environment variable contains the exporter name. This
+// will panic if name has already been registered.
+func RegisterSpanExporter(name string, e trace.SpanExporter) {
+	if err := envRegistry.store(name, e); err != nil {
+		// envRegistry.store will return errDupReg if name is already
+		// registered. Panic here so the user is made aware of the duplicate
+		// registration, which could be done by malicious code trying to
+		// intercept cross-cutting concerns.
+		//
+		// Panic for all other errors as well. At this point there should not
+		// be any other errors returned from the store operation. If there
+		// are, alert the developer that adding them as soon as possible that
+		// they need to be handled here.
+		panic(err)
+	}
+}
+
+// SpanExporter returns a SpanExporter using passed in name
+// using the list of registered SpanExporters. Each name must match an
+// already registered SpanExporter or a default (otlp).
+//
+// An error is returned for any unknown exporters.
+func SpanExporter(name string) (trace.SpanExporter, error) {
+	if name == "otlp" {
```

Comment:
Instead of having this default case, is it possible to initialize the registry with the grpc and http/protobuff otlp exporters during initialization?  Or does this logic end up being more efficient than storing otlp exporters in the registry?

Reply:
This does delay the exporter creation and therefor it's resources until it's selected as the exporter to be used which is nice. eg you won't end up with unused exporters in the registry.

Another benefit of lazily creating them like this is that we can test expected behaviour by setting env vars in unit tests before calling retrieving an exporter. We wouldn't get the opportunity to test the different exporter options if they were created on package initialisation, .

Reply:
I've updated the logic to register the OTLP exporter during `init()`. We only need one registration for an OTLP exporter, but under both `""` and `"otlp"` keys.